core/src/apps/cardano/sign_tx.py

```diff
@@ -743,31 +854,38 @@ async def _show_change_output_staking_warnings(
             )
 
 
-# addresses from the same account as inputs should be hidden
-def _should_hide_output(output: list[int], inputs: list[CardanoTxInputType]) -> bool:
-    for tx_input in inputs:
-        inp = tx_input.address_n
-        if (
-            len(output) != BIP_PATH_LENGTH
-            or output[ACCOUNT_PATH_INDEX] != inp[ACCOUNT_PATH_INDEX]
-            or output[ACCOUNT_PATH_INDEX] > MAX_SAFE_ACCOUNT_INDEX
-            or output[-2] >= 2
-            or output[-1] >= MAX_SAFE_CHANGE_ADDRESS_INDEX
-        ):
-            return False
+def _should_hide_output(path: list[int]) -> bool:
+    """Return whether the output address is from a safe path, so it could be hidden."""
```

Comment:
this means that any path from any account is considered safe, so in particular, transfer between my own accounts will have no visible output confirmation screen?

Reply:
i propose the following modification:

- if there are "internal" outputs (specified by address_params instead of address), at least one "external" output must exist, otherwise transaction is rejected at end of outputs stage
- all provided `address_n`s (for outputs AND witnesses) must match the same account, otherwise transaction is rejected at first non-matching entry

Reply:
Before `_should_hide_output` is called in `_show_output`, `address_n` is checked to match against `SCHEMA_ADDRESS` - if it doesn't match, signing will either fail or a warning will be shown - depending on `safety_checks.is_strict()`.

We have refactored `_show_output` in the multi-sig PR so I wouldn't want to touch it here if there are no serious security implications and if there are, I would rather deal with those in a separate PR. The `_should_hide_output` function has also been there before tx streaming. It's just been modified to not take inputs into consideration as that is not possible anymore and checking against the inputs didn't really add much guarantees anyways.

Given all this, I'm don't think the proposed modifications are necessary.

Reply:
The protection was not really strict before this, but at least there was guarantee that all change is going back into the same account. Now the only guarantee is that it matches `SCHEMA_ADDRESS` (fwiw you could basically replace the whole function by `return SCHEMA_ADDRESS.match(...)`).

This change basically throws out the account isolation model. Anything counts as "change output" as long as it's going into one of your accounts. Malicious software can deanonymize your spending by splitting up change into all your accounts, for example.

> not take inputs into consideration as that is not possible anymore

You can still match against witness paths, as proposed. (unless it is a known scenario to request witnesses from multiple accounts at the same time?)

> I would rather deal with those in a separate PR.

If it makes it easier for you, we can finalize this PR and commit it into a separate branch.

Reply:
Ok, I think I get your point.
> If it makes it easier for you, we can finalize this PR and commit it into a separate branch.

Let's do it here.

> if there are "internal" outputs (specified by address_params instead of address), at least one "external" output must exist, otherwise transaction is rejected at end of outputs stage

Why is this restriction needed?

I agree with the rest, although rather than failing, maybe showing a warning would be enough? Or failing/warning should depend on `safety_checks.is_strict()`.

Reply:
> > if there are "internal" outputs (specified by address_params instead of address), at least one "external" output must exist, otherwise transaction is rejected at end of outputs stage
> 
> Why is this restriction needed?

After thinking about it some more, it probably isn't. This is purely a UX problem: you'll be confirming a transaction with zero total, but that's basically correct information.

> rather than failing, maybe showing a warning would be enough? Or failing/warning should depend on `safety_checks.is_strict()`.

This should be a hard failure, because it's fully avoidable from the host side by sending a "correct" transaction: instead of multiple change outputs going to multiple accounts, specify them via addresses (that will be displayed for the user to confirm).

Reply:
Hm, it seems I'll have to consult the issue with IOHK - I'll get back to you ASAP.

Reply:
I'm not sure simply rejecting the transaction is a good idea. Until now the mixing of accounts was possible, although yes, the outputs would be shown. We might discreetly break some integrations if we decide to start failing from now on.

IMHO if we show a warning accompanying before outputs and witnesses which contain a different path than the first item containing a path (output or witness), it should be safe enough for the users. IMHO showing the outputs as address parameters + the address is better/less confusing than showing an encoded address.

> Malicious software can deanonymize your spending by splitting up change into all your accounts, for example.

Is deanonymization an issue if a malicious wallet can easily bulk export any of your public keys and do whatever with them? Or is bulk export limited to a single account?

Regardless of whether we end up failing or showing, we need to store the account of the first encountered path. Do you have a suggestion for storing this account?

One alternative is that a `transaction_account` variable would get passed around from the top level `sign_tx` function and it would also have to be returned from the functions which can modify it e.g. `process_outputs`, `process_certificates`, .... 

Another is to introduce a `CardanoTxContext`class which would store the `transaction_account` and also `network_id` and `protocol_magic`. The ugly thing is that `transaction_account` would be modified at some point during tx signing and not at the start.

Another would be to determine `transaction_account` in Connect and pass it into Trezor or let it be passed into Connect and forwarded to Trezor - which doesn't really seem right and it would perhaps unnecessarily change the protocol.

Reply:
> Is deanonymization an issue if a malicious wallet can easily bulk export any of your public keys and do whatever with them? Or is bulk export limited to a single account?

This way the Trezor itself signs a cryptographic proof that the accounts are linked; the best the user can claim is that the accounts are not theirs, but they deliberately sent them money. With bulk export, the attacker has no way of proving to a 3rd party that the XPUBs come from the same device.

Another issue which does not strictly fall under "security" is that an attacker can basically freely move funds between user's accounts.

> IMHO if we show a warning accompanying before outputs and witnesses which contain a different path

Mixing witness accounts, if accompanied by a warning, is OK (there's still some space left for a deanonymization attack, but at that point it's mostly theoretical) --- as long as _no outputs were hidden_ previously. The logic goes like this:
(a) either you're spending from a single account and sending back change to that same account, in which case change detection kicks in and the change output is not shown. (FWIW in Bitcoin we also show a warning when more than two change outputs are detected, because that indicates suspicious activity)
(b) or you're spending from multiple sources and/or sending to multiple accounts, in which case it's impossible to determine which outputs are "change", and all outputs must be shown.

I don't think "warning: some outputs were not displayed" is something that makes any sort of sense, so hard failure is the only option here.

We previously checked all outputs against _first input_, which at least ensures that the change is not scattered, but still doesn't necessarily display the right thing. I consider it a security-adjacent bug. If any existing integrations relied on this bug, I'm OK with breaking them by fixing the bug.

--- 

> Regardless of whether we end up failing or showing, we need to store the account of the first encountered path. Do you have a suggestion for storing this account?

Have a look at Bitcoin's matcher: https://github.com/trezor/trezor-firmware/blob/master/core/src/apps/bitcoin/sign_tx/matchcheck.py
It's not a full "transaction context", you would probably only need to pass it into "process outputs" and "process witnesses" and not any deeper, from a quick look.

(if applicable, you could copy-paste the code and we'd resolve extracting back common functionality in some other PR)

If we need to pass this deeper somewhere, i'd consider making the TxContext and placing a matcher as one of its properties.

Reply:
Alright, I'll implement the matchchecker today and so transaction signing would fail if multiple accounts are used. I think we also need to add certificates and withdrawals to the matchchecker though, because they require witnesses which are included along the other witness requests and we have no way to distinguish which came from where. This means that basically all the paths (change output payment path, certificate path, withdrawal path, input/witness path) in a transaction will have to belong to a single account. Do you agree?

Reply:
I'll have to go with your opinion here because i don't really know what certificates and withdrawals _do_ -- but yes, presumably the same logic (single account yes/no) would apply too.

Reply:
I've implemented the `AccountPathChecker`. All relevant paths are being included in it including pool owners as the witness for that is also being passed in from the outside. I had to change `SCHEMA_ADDRESS` to `SCHEMA_PAYMENT` in order to simplify the validation for hiding outputs (as I mentioned before, this has already been refactored in the mutli-sig PR anyways so I wouldn't want to further modify it here if not necessary). I had to modify some tests which didn't follow the single account rule and I added more tests which verify that transactions with mixed accounts indeed do fail.

Reply:
ACK on the matcher usage, but I believe the "Byron to Shelley" scenario must be supported -- isn't that what a past issue was about? Trezor wrongly showing a change output in case Byron and Shelley inputs are mixed?

At least some Cardano wallets consider Byron and Shelley accounts to be the same account. (I presume it makes more sense in Cardano than in Bitcoin, if Shelley is basically a "rolling upgrade"? But you'd need to confirm with IOHK, i'm the less knowledgeable party here)

So presumably to fix that, instead of picking `to_account_path` as the matcher value, you'd just pick `path[ACCOUNT_PATH_INDEX]` ?

Reply:
We've confirmed with IOHK that Daedalus and Yoroi that they don't combine the two wallets. I.e. if there are Byron inputs then Shelley outputs are external e.g. migrating the whole wallet from Byron to Shelley. However, it seems Adalite can combine them in some instances. But IMHO it doesn't matter if we're talking about `44'/1815'/0'` vs `1852'/1815'/0'` or `1852'/1815'/0'` vs `1852'/1815'/1'`. The same issues arise. Since Daedalus and Yoroi don't mix the two wallets, the user wouldn't see the funds potentially sent to the `44'/...` (hidden) change output. We'll discuss this internally ASAP. Sorry this is taking so long.

Reply:
> he same issues arise. Since Daedalus and Yoroi don't mix the two wallets, the user wouldn't see the funds potentially sent to the `44'/...` (hidden) change output.

This is a good point. Thanks for confirming. I'm OK with keeping the implementation as-is if you agree.